app/src/org/commcare/android/database/HybridFileBackedSqlHelpers.java

```diff
@@ -0,0 +1,169 @@
+            throw new UserStorageClosedException(e.getMessage());
+        }
+
+        String[] columns = new String[]{DatabaseHelper.FILE_COL};
+        c = db.query(table, columns, DatabaseHelper.ID_COL + "=?",
+                new String[]{String.valueOf(id)}, null, null, null);
+
+        try {
+            c.moveToFirst();
+            return c.getString(c.getColumnIndexOrThrow(DatabaseHelper.FILE_COL));
+        } finally {
+            if (c != null) {
+                c.close();
+            }
+        }
+    }
+
+    protected static List<String> getFilesToRemove(List<Integer> idsBeingRemoved,
+                                                   AndroidDbHelper helper,
+                                                   String table) {
+        ArrayList<String> files = new ArrayList<>();
+        // delete files storing data for entries being removed
+        for (Integer id : idsBeingRemoved) {
+            String filename = HybridFileBackedSqlHelpers.getEntryFilename(helper, table, id);
+            if (filename != null && new File(filename).exists()) {
+                files.add(filename);
+            }
+        }
+        return files;
+    }
+
+    protected static void removeFiles(List<String> filesToRemove) {
+        for (String filename : filesToRemove) {
+            File datafile = new File(filename);
+            datafile.delete();
+        }
+    }
+
+    protected static void setFileAsOrphan(SQLiteDatabase db, String filename) {
+        db.beginTransaction();
+        try {
+            ContentValues cv = new ContentValues();
+            cv.put(DatabaseHelper.FILE_COL, filename);
+            db.insertOrThrow(DbUtil.orphanFileTableName, DatabaseHelper.FILE_COL, cv);
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    protected static void unsetFileAsOrphan(SQLiteDatabase db, String filename) {
+        int deleteCount = db.delete(DbUtil.orphanFileTableName, DatabaseHelper.FILE_COL + "=?", new String[]{filename});
+        if (deleteCount != 1) {
+            Logger.log(AndroidLogger.SOFT_ASSERT,
+                    "Unable to unset orphaned file: " + deleteCount + " entries effected.b");
+        }
+    }
+
+    public static void removeOrphanedFiles(SQLiteDatabase db) {
+        Cursor cur = db.query(DbUtil.orphanFileTableName, new String[] {DatabaseHelper.FILE_COL}, null, null, null, null, null);
+        ArrayList<String> files = new ArrayList<>();
+        try {
+            if (cur.getCount() > 0) {
+                cur.moveToFirst();
+                int fileColIndex = cur.getColumnIndexOrThrow(DatabaseHelper.FILE_COL);
+                while (!cur.isAfterLast()) {
+                    files.add(cur.getString(fileColIndex));
+                    cur.moveToNext();
+                }
+            }
+        } finally {
+            if (cur != null) {
+                cur.close();
+            }
+        }
+
+        removeFiles(files);
```

Comment:
We're basically assuming that the filenames are globally unique, but if not, this approach can result in weird behavior like

Table[A,B]

Delete all files: A succeeds, B fails, File System State: B, Table State[A,B]

App checks whether 'A' exists on filesystem before putting something there, sees that it doesn't, adds a new file to slot A

purge triggered

"New" A is removed.

Seems like a fine tradeoff, but it seems worth being explicit that this helper system only works with globally unique names